File: apps/mobile/ios/Runner/video_metadata_extractor/VideoMetadataMethodChannelHandler.swift
import Foundation
import Flutter

/// Handles Flutter MethodChannel calls for VideoMetadataExtractor operations.
/// Equivalent to Android's VideoMetadataMethodChannelHandler.
class VideoMetadataMethodChannelHandler: NSObject {
    
    private static let channelName = "video_helper"
    
    private let videoMetadataExtractor: VideoMetadataExtractor
    private let methodChannel: FlutterMethodChannel
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        self.videoMetadataExtractor = VideoMetadataExtractorImpl()
        self.methodChannel = FlutterMethodChannel(
            name: VideoMetadataMethodChannelHandler.channelName,
            binaryMessenger: binaryMessenger
        )
        
        super.init()
        
        methodChannel.setMethodCallHandler { [weak self] call, result in
            self?.handleMethodCall(call: call, result: result)
        }
    }
    
    private func handleMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "extractThumbnail":
            handleExtractThumbnail(call: call, result: result)
        case "getVideoDuration":
            handleGetVideoDuration(call: call, result: result)
        case "getVideoMetadata":
            handleGetVideoMetadata(call: call, result: result)
        case "openVideoInViewer":
            handleOpenVideoInViewer(call: call, result: result)
        case "saveVideoToGallery":
            handleSaveVideoToGallery(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - Method Handlers
    
    private func handleExtractThumbnail(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let urlString = args["uri"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "URI is null", details: nil))
            return
        }
        
        let timeMs = (args["timeMs"] as? NSNumber)?.int64Value ?? 0
        
        // Run on background thread to avoid blocking UI
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            let thumbnailPath = self?.videoMetadataExtractor.extractThumbnail(
                urlString: urlString,
                timeMs: timeMs
            )
            
            DispatchQueue.main.async {
                result(thumbnailPath)
            }
        }
    }
    
    private func handleGetVideoDuration(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let urlString = args["uri"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "URI is null", details: nil))
            return
        }
        
        // Run on background thread to avoid blocking UI
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            let durationSeconds = self?.videoMetadataExtractor.getVideoDuration(urlString: urlString)
            
            DispatchQueue.main.async {
                result(durationSeconds)
            }
        }
    }
    
    private func handleGetVideoMetadata(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let urlString = args["uri"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "URI is null", details: nil))
            return
        }
        
        // Run on background thread to avoid blocking UI
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            let metadata = self?.videoMetadataExtractor.getVideoMetadata(urlString: urlString)
            
            DispatchQueue.main.async {
                result(metadata)
            }
        }
    }
    
    private func handleOpenVideoInViewer(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let urlString = args["uri"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "URI is null", details: nil))
            return
        }
        
        let success = videoMetadataExtractor.openVideoInViewer(urlString: urlString)
        result(success)
    }
    
    private func handleSaveVideoToGallery(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let sourcePath = args["sourcePath"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Source path is null", details: nil))
            return
        }
        
        // Run on background thread as photo library operations can be slow
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            let success = self?.videoMetadataExtractor.saveVideoToGallery(sourcePath: sourcePath) ?? false
            
            DispatchQueue.main.async {
                result(success)
            }
        }
    }
}

File: apps/mobile/android/app/src/main/kotlin/open/fpvlabs/stera/video_metadata_extractor/VideoMetadataMethodChannelHandler.kt
package open.fpvlabs.stera.video_metadata_extractor

import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

/**
 * Handles Flutter MethodChannel calls for VideoMetadataExtractor operations.
 */
class VideoMetadataMethodChannelHandler(
    context: Context,
    binaryMessenger: BinaryMessenger
) {
    companion object {
        private const val CHANNEL_NAME = "video_helper"
    }

    private val videoMetadataExtractor: VideoMetadataExtractor = VideoMetadataExtractorImpl(context)
    private val methodChannel: MethodChannel = MethodChannel(binaryMessenger, CHANNEL_NAME)

    init {
        methodChannel.setMethodCallHandler { call, result ->
            when (call.method) {
                "extractThumbnail" -> handleExtractThumbnail(call, result)
                "getVideoDuration" -> handleGetVideoDuration(call, result)
                "openVideoInViewer" -> handleOpenVideoInViewer(call, result)
                "saveVideoToGallery" -> handleSaveVideoToGallery(call, result)
                else -> result.notImplemented()
            }
        }
    }

    private fun handleExtractThumbnail(call: io.flutter.plugin.common.MethodCall, result: MethodChannel.Result) {
        val uriString = call.argument<String>("uri")
        val timeMs = call.argument<Number>("timeMs") ?: 0

        if (uriString == null) {
            result.error("INVALID_ARGUMENT", "URI is null", null)
            return
        }

        // Run on background thread to avoid blocking Android main thread
        // Flutter side also uses isolate for double protection
        Thread {
            try {
                val thumbnailPath = videoMetadataExtractor.extractThumbnail(uriString, timeMs.toLong())
                result.success(thumbnailPath)
            } catch (e: Exception) {
                result.error("EXTRACTION_ERROR", e.message, null)
            }
        }.start()
    }

    private fun handleGetVideoDuration(call: io.flutter.plugin.common.MethodCall, result: MethodChannel.Result) {
        val uriString = call.argument<String>("uri")

        if (uriString == null) {
            result.error("INVALID_ARGUMENT", "URI is null", null)
            return
        }

        // Run on background thread to avoid blocking Android main thread
        // Flutter side also uses isolate for double protection
        Thread {
            try {
                val durationSeconds = videoMetadataExtractor.getVideoDuration(uriString)
                result.success(durationSeconds)
            } catch (e: Exception) {
                result.error("DURATION_ERROR", e.message, null)
            }
        }.start()
    }

    private fun handleOpenVideoInViewer(call: io.flutter.plugin.common.MethodCall, result: MethodChannel.Result) {
        val uriString = call.argument<String>("uri")

        if (uriString != null) {
            val success = videoMetadataExtractor.openVideoInViewer(uriString)
            result.success(success)
        } else {
            result.error("INVALID_ARGUMENT", "URI is null", null)
        }
    }

    private fun handleSaveVideoToGallery(call: io.flutter.plugin.common.MethodCall, result: MethodChannel.Result) {
        val sourcePath = call.argument<String>("sourcePath")

        if (sourcePath != null) {
            val success = videoMetadataExtractor.saveVideoToGallery(sourcePath)
            result.success(success)
        } else {
            result.error("INVALID_ARGUMENT", "Source path is null", null)
        }
    }
}
